### PR TITLE
Improve partial unification handling

### PR DESCRIPTION
**Pull request by smarter**
Currently failing because it breaks shapeless.

### CODE REVIEW

**Comment by smarter** [comment]
@milessabin This PR broke shapeless compilation as described in https://github.com/lampepfl/dotty/pull/9523/commits/9e7cc103d9ff184b57ab0e812fcaed49e7d8f734, do you have the time to try to diagnose/minimize what's going on here?

**Comment by milessabin** [comment]
It looks like the sort of change which would be likely to create new ambiguities and I can imagine that breaking the `Pure` derivation.

I'm afraid I'm really not able to commit the time to look into this properly right now :-(

**Comment by smarter** [comment]
Understood, thanks for taking a look. I'll look into this a bit more but since I'd like to get this in sooner rather than later (since it's a net improvement, it's potentially disruptive, and it's exercised by cats), I will probably have to comment out some of the code related to Pure in the branch of shapeless in our community build.

**Comment by smarter** [comment]
test performance please

**Comment by smarter** [comment]
Good news! I was able to find the root cause of the issues i was seeing with shapeless: https://github.com/lampepfl/dotty/pull/9523/commits/fb7bc51eb85f46f537a07b0305622a86057e20ad, now everything is green! :sparkles:  (the diverging implicit issue is still concerning, but it seems to be a manifestation of something already present on master: https://github.com/lampepfl/dotty/issues/9504)

**Comment by smarter** [comment]
> the diverging implicit issue is still concerning, but it seems to be a manifestation of something already present on master: #9504

#9504 is fixed, but the implicit issue is still here, however the issue goes away if the parameter is not marked by-name now (it didn't before), and I've been able to find a similar issue on master: https://github.com/lampepfl/dotty/issues/9568

**Comment by odersky** [comment]
#9568 is fixed by #9575

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -849,6 +849,85 @@ class TypeComparer(using val comparerCtx: Context) extends ConstraintHandling wi
...
+     *
+     *    [T_0, ..., T_k-1] =>> otherTycon[otherArgs(0), ..., otherArgs(d-1), T_0, ..., T_k-1]
+     *
+     *  where `T_n` has the same bounds as `otherTycon.typeParams(d+n)`
+     *
+     *  Historical note: this strategy is known in Scala as "partial unification"
+     *  (even though the type constructor variable isn't actually unified but only
+     *  has one of its bounds constrained), for background see:
+     *  - The infamous SI-2712: https://github.com/scala/bug/issues/2712
+     *  - The PR against Scala 2.12 implementing -Ypartial-unification: https://github.com/scala/scala/pull/5102
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:899]: Nice explanation!

**Pull request merged**